File: android_ChirpTransmitter/app/src/main/java/com/example/chirptransmitter/utils/ChirpGenerator.java
package com.example.chirptransmitter.utils;

/**
 * // only play sound on left
 * for(int i = 0; i < count; i += 2){
 *     short sample = (short)(Math.sin(2 * Math.PI * i / (44100.0 / freqHz)) * 0x7FFF);
 *     samples[i + 0] = sample;
 *     samples[i + 1] = 0;
 * }
 * // only play sound on right
 * for(int i = 0; i < count; i += 2){
 *     short sample = (short)(Math.sin(2 * Math.PI * i / (44100.0 / freqHz)) * 0x7FFF);
 *     samples[i + 0] = 0;
 *     samples[i + 1] = sample;
 * }
 *
 * */




public class ChirpGenerator {

    public final static int SPEAKER_TYPE_ZERO   = 0;
    /** Using top speaker only */
    public final static int SPEAKER_TYPE_TOP    = 1;
    /** Using bottom speaker only */
    public final static int SPEAKER_TYPE_BOTTOM = 2;
    /** Using top and bottom speaker */
    public final static int SPEAKER_TYPE_BOTH   = 3;


    static public byte[] genPCM16SingleToneBytes(
            int sampleRate,
            double freq,
            double tSweep,
            double tFrame,
            int times,
            boolean applyHamming) {


        int chirpSampleCnt = (int) Math.round(tSweep * sampleRate);
        int frameSampleCnt = (int) Math.round(tFrame * sampleRate);

        // generate Window
        int windowSize = chirpSampleCnt;

        double[] sample_top = new double[frameSampleCnt * times];
        double[] sample_bottom = new double[frameSampleCnt * times];

        double[] factors = new double[windowSize];
        for(int i = 0; i < windowSize; i++) {
            if (applyHamming) {
                factors[i] = 0.5 * (1 - Math.cos((2 * Math.PI * i) / (windowSize-1)));
            } else {
                factors[i] = 1.0d;
            }
        }

        int t = 0;
        while (t < times) {
            double instfreq_top, instfreq_bottom, numerator;
            for (int i = 0; i < chirpSampleCnt; i++) {
                numerator = (double)(i) / (double) chirpSampleCnt;
                instfreq_top = numerator * freq; //fStart + (0.5 * numerator * (fHalf - fStart));
                sample_top[i + t * frameSampleCnt] = factors[i] * Math.cos(2.0 * Math.PI * i / (sampleRate / instfreq_top));
            }
            for (int i = chirpSampleCnt; i < frameSampleCnt; i++) {
                sample_top[i + t * frameSampleCnt] = 0.0;
            }

            for (int i = 0; i < chirpSampleCnt; i++) {
                numerator = (double)(i) / (double) chirpSampleCnt;
                instfreq_bottom = numerator * freq; //fHalf + (0.5 * numerator * (fEnd - fHalf));
                sample_bottom[i + t * frameSampleCnt] = factors[i] * Math.cos(2.0 * Math.PI * i / (sampleRate / instfreq_bottom ));
            }
            for (int i = chirpSampleCnt; i < frameSampleCnt; i++) {
                sample_bottom[i + t * frameSampleCnt] = 0.0;
            }

            t++;
        }

        int idx = 0;
        byte[] data = new byte[2 * 2 * frameSampleCnt * times];
        for (final double dVal : sample_top) {
            // left channel data
            final short val = (short) ((dVal * 32767));
            data[idx++] = (byte) (val & 0x00ff);
            data[idx++] = (byte) ((val & 0xff00) >>> 8);

            // right channel data
            data[idx++] = (byte) (0);
            data[idx++] = (byte) (0);
        }

//        idx = 0; // reset idx
//        for (final double dVal : sample_bottom) {
//            // left channel data
//            idx++;
//            idx++;
//
//            // right channel data
//            final short val = (short) ((dVal * 32767));
//            data[idx++] = (byte) (val & 0x00ff);
//            data[idx++] = (byte) ((val & 0xff00) >>> 8);
//        }

        return data;
    }


    /** Generate the speaker chirp signal data */
    static public byte[] genPCM16StereoBytes(
            int sampleRate,
            double fStart,
            double fEnd,
            double tSweep,
            double tFrame,
            int times,
            boolean applyHamming,
            int speakerType) {

        double fHalf = (fEnd + fStart) / 2.0;

        int chirpSampleCnt = (int) Math.round(tSweep * sampleRate);
        int frameSampleCnt = (int) Math.round(tFrame * sampleRate);

        // generate Window
        int windowSize = chirpSampleCnt;

        double[] sample_top = new double[frameSampleCnt * times];
        double[] sample_bottom = new double[frameSampleCnt * times];

        double[] factors = new double[windowSize];
        for(int i = 0; i < windowSize; i++) {
            if (applyHamming) {
                factors[i] = 0.5 * (1 - Math.cos((2 * Math.PI * i) / (windowSize-1)));
            } else {
                factors[i] = 1.0d;
            }
        }

        int t = 0;
        while (t < times) {
            double instfreq_top, instfreq_bottom, numerator;
            for (int i = 0; i < chirpSampleCnt; i++) {
                numerator = (double)(i) / (double) chirpSampleCnt;
                instfreq_top = fStart + (0.5 * numerator * (fHalf - fStart));
                sample_top[i + t * frameSampleCnt] = factors[i] * Math.cos(2.0 * Math.PI * i / (sampleRate / instfreq_top));
            }
            for (int i = chirpSampleCnt; i < frameSampleCnt; i++) {
                sample_top[i + t * frameSampleCnt] = 0.0;
            }

            for (int i = 0; i < chirpSampleCnt; i++) {
                numerator = (double)(i) / (double) chirpSampleCnt;
                instfreq_bottom = fHalf + (0.5 * numerator * (fEnd - fHalf));
                sample_bottom[i + t * frameSampleCnt] = factors[i] * Math.cos(2.0 * Math.PI * i / (sampleRate / instfreq_bottom));
            }
            for (int i = chirpSampleCnt; i < frameSampleCnt; i++) {
                sample_bottom[i + t * frameSampleCnt] = 0.0;
            }

            t++;
        }

        int idx = 0;
        byte[] data = new byte[2 * 2 * frameSampleCnt * times];
        for (final double dVal : sample_top) {
            // left channel data
            final short val = (short) ((dVal * 32767));
            data[idx++] = (byte) (val & 0x00ff);
            data[idx++] = (byte) ((val & 0xff00) >>> 8);

            // right channel data
            data[idx++] = (byte) (0);
            data[idx++] = (byte) (0);
        }

        idx = 0; // reset idx
        for (final double dVal : sample_bottom) {
            // left channel data
            idx++;
            idx++;

            // right channel data
            final short val = (short) ((dVal * 32767));
            data[idx++] = (byte) (val & 0x00ff);
            data[idx++] = (byte) ((val & 0xff00) >>> 8);
        }

        return data;
    }


    /** Generate the speaker chirp signal data */
    static public byte[] genPCM16MonoToneBytes(
            int sampleRate,
            double fStart,
            double fEnd,
            double tSweep,
            double tFrame,
            int times,
            boolean applyHamming,
            int speakerType) {

        int chirpSampleCnt = (int) Math.round(tSweep * sampleRate);
        int frameSampleCnt = (int) Math.round(tFrame * sampleRate);

        // generate Window
        int windowSize = chirpSampleCnt;
        double[] sample = new double[frameSampleCnt * times];
        double[] factors = new double[windowSize];
        for(int i = 0; i < windowSize; i++) {
            if (applyHamming) {
                factors[i] = 0.5 * (1 - Math.cos((2 * Math.PI * i) / (windowSize-1)));
            } else {
                factors[i] = 1.0d;
            }
        }

        int t = 0;
        while (t < times) {
            double instfreq, numerator;
            for (int i = 0; i < chirpSampleCnt; i++)
            {
                numerator = (double)(i) / (double) chirpSampleCnt;
                instfreq = fStart + (0.5 * numerator * (fEnd - fStart));
                // apply window
                sample[i + t * frameSampleCnt] = factors[i] * Math.cos(2.0 * Math.PI * i / (sampleRate / instfreq));
            }
            for (int i = chirpSampleCnt; i < frameSampleCnt; i++) {
                sample[i + t * frameSampleCnt] = 0.0;
            }
            t++;
        }

        int idx = 0;
        byte[] data = new byte[2 * 2 * frameSampleCnt * times];
        for (final double dVal : sample) {

            if (speakerType == SPEAKER_TYPE_BOTTOM){
                // left channel data
                data[idx++] = (byte) (0);
                data[idx++] = (byte) (0);

                // right channel data
                final short val = (short) ((dVal * 32767)); //scale to maximum amplitude
                data[idx++] = (byte) (val & 0x00ff);
                data[idx++] = (byte) ((val & 0xff00) >>> 8);
            }else if (speakerType == SPEAKER_TYPE_TOP){
                // left channel data
                final short val = (short) ((dVal * 32767));
                data[idx++] = (byte) (val & 0x00ff);
                data[idx++] = (byte) ((val & 0xff00) >>> 8);

                // right channel data
                data[idx++] = (byte) (0);
                data[idx++] = (byte) (0);
            }
            else if (speakerType == SPEAKER_TYPE_BOTH) {
                // left channel data
                final short val = (short) ((dVal * 32767));
                data[idx++] = (byte) (val & 0x00ff);
                data[idx++] = (byte) ((val & 0xff00) >>> 8);

                // right channel data
                data[idx++] = (byte) (val & 0x00ff);
                data[idx++] = (byte) ((val & 0xff00) >>> 8);

            }
            else if (speakerType == SPEAKER_TYPE_ZERO) {
                data[idx++] = (byte) (0);
                data[idx++] = (byte) (0);

                data[idx++] = (byte) (0);
                data[idx++] = (byte) (0);

            }
            else {
                data[idx++] = (byte) (0);
                data[idx++] = (byte) (0);

                data[idx++] = (byte) (0);
                data[idx++] = (byte) (0);
                break;
            }

        }

        return data;
    }

}

File: UltrasoundPFT/app/src/main/cpp/tools/methods/chirp_generator.cpp
//
// Created by Dell on 2024/2/23.
//
#include <vector>
#include <cmath>

#include "chirp_generator.h"


std::vector<double> genPCM16MonoToneBytes(ChirpParameters params) {

    int chirpSampleCnt = std::round(params.chirpDuration * params.sampleRate);
    int frameSampleCnt = std::round(params.chirpDuration * params.sampleRate);

    // 生成窗函数
    std::vector<double> factors;
    factors.reserve(chirpSampleCnt);
    for (int i = 0; i < chirpSampleCnt; i++) {
        if (params.applyHamming) {
            factors.push_back(0.5 * (1 - std::cos((2 * M_PI * i) / (chirpSampleCnt - 1))));
        } else {
            factors.push_back(1.0);
        }
    }

    std::vector<double> sample;
    sample.reserve(frameSampleCnt * params.chirpCircle);

    int t = 0;
    while (t < params.chirpCircle) {
        double instfreq, numerator;
        for (int i = 0; i < chirpSampleCnt; i++) {
            numerator = static_cast<double>(i) / chirpSampleCnt;
            instfreq = params.chirpStartFrq +
                    (0.5 * numerator * (params.chirpStopFrq - params.chirpStartFrq) );

            // apply windows function
            double value = factors[i] * std::cos(2.0 * M_PI * i / (params.sampleRate / instfreq));
            sample.push_back(value);
        }
        for (int i = chirpSampleCnt; i < frameSampleCnt; i++) {
            sample.push_back(0.0);
        }
        t++;
    }

    std::vector<double> data;
    data.reserve(frameSampleCnt * params.chirpCircle);

    int idx = 0;
    for (const double& dVal : sample) {
//        int16_t val = static_cast<int16_t>(dVal * 32767);
//        data.push_back(val);
        data.push_back(dVal);
    }

    return data;
}



std::vector<double> genPCM16MonoToneBytes_Sin(ChirpParameters params) {

    int chirpSampleCnt = std::round(params.chirpDuration * params.sampleRate);
    int frameSampleCnt = std::round(params.chirpDuration * params.sampleRate);

    // 生成窗函数
    std::vector<double> factors;
    factors.reserve(chirpSampleCnt);
    for (int i = 0; i < chirpSampleCnt; i++) {
        if (params.applyHamming) {
            factors.push_back(0.5 * (1 - std::sin((2 * M_PI * i) / (chirpSampleCnt - 1))));
        } else {
            factors.push_back(1.0);
        }
    }

    std::vector<double> sample;
    sample.reserve(frameSampleCnt * params.chirpCircle);

    int t = 0;
    while (t < params.chirpCircle) {
        double instfreq, numerator;
        for (int i = 0; i < chirpSampleCnt; i++) {
            numerator = static_cast<double>(i) / chirpSampleCnt;
            instfreq = params.chirpStartFrq +
                       (0.5 * numerator * (params.chirpStopFrq - params.chirpStartFrq) );

            // apply windows function
            double value = factors[i] * std::sin(2.0 * M_PI * i / (params.sampleRate / instfreq));
            sample.push_back(value);
        }
        for (int i = chirpSampleCnt; i < frameSampleCnt; i++) {
            sample.push_back(0.0);
        }
        t++;
    }

    std::vector<double> data;
    data.reserve(frameSampleCnt * params.chirpCircle);

    int idx = 0;
    for (const double& dVal : sample) {
//        int16_t val = static_cast<int16_t>(dVal * 32767);
//        data.push_back(val);
        data.push_back(dVal);
    }

    return data;
}


std::vector<double> cosToSin(const std::vector<double>& cos_sequence) {
    std::vector<double> sin_sequence;
    sin_sequence.reserve(cos_sequence.size());

    for (double cos_val : cos_sequence) {
        double sin_val = std::sin(M_PI / 2 - std::acos(cos_val));
        sin_sequence.push_back(sin_val);
    }

    return sin_sequence;
}


